src/compiler/checker.ts

```diff
@@ -6134,10 +6136,23 @@ module ts {
                     }
                     lastSymbol = symbol;
 
-                    for (var j = result.length; j > pos; j--) {
+                    // specialized signatures always need to be placed before non-specialized signatures regardless
+                    // of the cutoff position; see GH#1133
+                    if (signature.hasStringLiterals) {
+                        splicePos = ++specializedPos;
+                        // The cutoff position needs to be increased to account for the fact that we are adding things
+                        // before the cutoff point. If the cutoff position is not incremented, merged interfaces will
+                        // start adding their merged signatures at the wrong position
+                        ++cutoffPos;

```

Comment:
What if the cutoffPos is less than specializedPos? Then it should not incremented, right?

Actually, it should, because if it is not, the following would not work:
```ts
interface A {
   (x: "hi"): string;
}
interface A {
   (x: string): string;
}
```
Here, even though the cutoffPos was before the specializedPos, it needs to be incremented, or else the general overload will jump above the specialized one.

Can you restate the comment to explain this? The key point is that there is an invariant that cutoffPos is always greater than or equal to specializedPos so that nothing is added before the specialized signatures.